File: src/contexts/core/kanji/domain/valueobjects/kanji-onyomi-reading.ts
import { ValueObjectValidationError } from "../../../../shared/errors/domain/ValueObjectValidationError";

class KanjiOnyomiReading {
    constructor(
        private readonly value: string,
    ) {
        if (value === '') {
            throw new ValueObjectValidationError('KanjiKunyomiReading should not be empty');
        }
    }

    get val () {
        return this.value;
    }
}

export { KanjiOnyomiReading };

File: libs/kernel/src/kanjis/domain/entities/kanji_onyomi_reading.rs
use crate::shared::domain::errors::DomainError;
use std::hash::{Hash, Hasher};

#[derive(Debug)]
pub struct KanjiOnyomiReading {
    value: String,
}

impl KanjiOnyomiReading {
    pub fn new(value: String) -> Result<Self, DomainError> {
        Ok(Self { value })
    }

    pub fn value(&self) -> String {
        self.value.clone()
    }
}

impl Clone for KanjiOnyomiReading {
    fn clone(&self) -> Self {
        Self::new(self.value.clone()).unwrap()
    }
}

impl PartialEq for KanjiOnyomiReading {
    fn eq(&self, other: &Self) -> bool {
        self.value() == other.value()
    }
}

impl Eq for KanjiOnyomiReading {}

impl Hash for KanjiOnyomiReading {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}
